File: ios/ReelVaultiOS/Views/SyncSetupSheet.swift
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 ReelVault Contributors

import SwiftUI
import ReelVaultKit

/// Sheet for configuring and running a one-off catalog sync operation.
///
/// Presented from LibraryGridScreen when the user taps "Sync to Remote" or
/// "Sync from Remote" in multi-select mode (both require canSync == true, i.e.
/// local embedded core is running AND a paired remote endpoint is stored).
///
/// Layout rules:
/// - NEVER wrap a NavigationStack in a VStack (breaks push transition).
/// - NEVER apply .safeAreaInset at window level (floats over UIKit nav bar).
/// - Progress banners use .safeAreaInset on content INSIDE the NavigationStack.
struct SyncSetupSheet: View {
    let direction: SyncDirection
    @Environment(\.dismiss) private var dismiss
    @EnvironmentObject private var router: AppRouter

    @State private var targetHeight: Int32 = 1080
    @State private var isRunning = false
    @State private var runResult: SyncRunResult?
    @State private var errorMessage: String?
    @State private var selectedTagId: String = ""
    @State private var selectedCollectionId: String = ""
    @State private var filterMinRating: Int32 = 0
    @State private var filterColorLabel: String = ""
    @State private var availableTags: [Tag] = []
    @State private var availableCollections: [Collection] = []

    var body: some View {
        NavigationStack {
            Form {
                Section("Sync Direction") {
                    Label(
                        direction == .toRemote
                            ? "Push local videos to remote server"
                            : "Pull remote videos to this device",
                        systemImage: direction == .toRemote
                            ? "arrow.up.to.line.circle"
                            : "arrow.down.to.line.circle"
                    )
                    .foregroundStyle(.secondary)
                }

                if direction == .fromRemote {
                    Section("Quality") {
                        Picker("Target Height", selection: $targetHeight) {
                            Text("720p").tag(Int32(720))
                            Text("1080p (Recommended)").tag(Int32(1080))
                            Text("Original").tag(Int32(0))
                        }
                        .pickerStyle(.menu)
                    }
                }

                Section("Filter (optional)") {
                    Picker("Tag", selection: $selectedTagId) {
                        Text("Any tag").tag("")
                        ForEach(availableTags) { tag in
                            Text(tag.name).tag(tag.id)
                        }
                    }
                    Picker("Collection", selection: $selectedCollectionId) {
                        Text("Any collection").tag("")
                        ForEach(availableCollections) { coll in
                            Text(coll.name).tag(coll.id)
                        }
                    }
                    Picker("Min Rating", selection: $filterMinRating) {
                        Text("Any rating").tag(Int32(0))
                        Text("★ or better").tag(Int32(1))
                        Text("★★ or better").tag(Int32(2))
                        Text("★★★ or better").tag(Int32(3))
                        Text("★★★★ or better").tag(Int32(4))
                        Text("★★★★★ only").tag(Int32(5))
                    }
                    Picker("Color Label", selection: $filterColorLabel) {
                        Text("Any color").tag("")
                        Text("Red").tag("red")
                        Text("Yellow").tag("yellow")
                        Text("Green").tag("green")
                        Text("Blue").tag("blue")
                        Text("Purple").tag("purple")
                    }
                }

                if isRunning {
                    Section("Progress") {
                        HStack(spacing: 12) {
                            ProgressView()
                            Text("Syncing…")
                                .foregroundStyle(.secondary)
                        }
                    }
                }

                if let result = runResult {
                    Section("Results") {
                        Label("\(result.completed) of \(result.total) synced", systemImage: "checkmark.circle.fill")
                            .foregroundStyle(.green)
                        if result.failed > 0 {
                            Label("\(result.failed) failed", systemImage: "exclamationmark.triangle.fill")
                                .foregroundStyle(.red)
                        }
                        if result.conflicts > 0 {
                            Label("\(result.conflicts) conflicts", systemImage: "arrow.triangle.branch")
                                .foregroundStyle(.orange)
                        }
                    }

                    if !result.errors.isEmpty {
                        Section("Errors") {
                            ForEach(result.errors, id: \.self) { err in
                                Text(err)
                                    .font(.caption)
                                    .foregroundStyle(.red)
                            }
                        }
                    }
                }

                if let msg = errorMessage {
                    Section {
                        Label(msg, systemImage: "xmark.octagon.fill")
                            .foregroundStyle(.red)
                    }
                }
            }
            .task { await loadFilterOptions() }
            .navigationTitle(direction == .toRemote ? "Sync to Remote" : "Sync from Remote")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                        .disabled(isRunning)
                }
                ToolbarItem(placement: .confirmationAction) {
                    if runResult != nil {
                        Button("Done") { dismiss() }
                    } else {
                        Button("Start Sync") {
                            Task { await startSync() }
                        }
                        .disabled(isRunning)
                    }
                }
            }
        }
    }

    // MARK: - Private

    private func startSync() async {
        guard let endpoint = router.lastPairedUploadEndpoint() else {
            errorMessage = "No paired server found. Pair a server first."
            return
        }
        isRunning = true
        errorMessage = nil
        defer { isRunning = false }

        // Build a quick-sync profile (no saved profile required for the one-shot UI).
        let profile = SyncProfile(
            name: "Quick Sync",
            peerKey: endpoint.fingerprintHex ?? "",
            direction: direction,
            filterJson: buildFilterJson(),
            targetHeight: targetHeight
        )

        // SyncManager requires the loopback port from the running local core.
        // LocalCore.port is set when startLocal() boots the embedded core.
        guard let localPort = LocalCore.port else {
            errorMessage = "Local core is not running."
            return
        }

        let manager = SyncManager(
            localPort: localPort,
            remoteHost: endpoint.host,
            remotePort: endpoint.mediaPort,   // gRPC port is mediaPort by convention
            remoteMediaPort: endpoint.mediaPort + 1,
            token: endpoint.bearerToken ?? "",
            fingerprint: endpoint.fingerprintHex ?? ""
        )

        await manager.startSync(profile: profile)
        runResult = manager.lastResult
    }

    private func buildFilterJson() -> String {
        var dict: [String: Any] = [:]
        if !selectedTagId.isEmpty { dict["filterTags"] = [selectedTagId] }
        if !selectedCollectionId.isEmpty { dict["collectionId"] = selectedCollectionId }
        if filterMinRating > 0 { dict["filterMinRating"] = Int(filterMinRating) }
        if !filterColorLabel.isEmpty { dict["filterColorLabel"] = filterColorLabel }
        guard !dict.isEmpty,
              let data = try? JSONSerialization.data(withJSONObject: dict, options: .sortedKeys),
              let str = String(data: data, encoding: .utf8)
        else { return "" }
        return str
    }

    private func loadFilterOptions() async {
        if let tags = try? await VideoRepository.shared.listTags() {
            availableTags = tags.sorted { $0.name < $1.name }
        }
        if let colls = try? await VideoRepository.shared.listCollections() {
            availableCollections = colls.filter { !$0.isSmart }.sorted { $0.name < $1.name }
        }
    }
}

#if DEBUG
#Preview {
    SyncSetupSheet(direction: .fromRemote)
        .environmentObject(AppRouter())
}
#endif

File: android/src/main/kotlin/com/reelvault/android/ui/screens/SyncSetupSheet.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 ReelVault Contributors

@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package com.reelvault.android.ui.screens

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.reelvault.android.ReelVaultApp
import com.reelvault.android.core.LocalCore
import com.reelvault.android.core.ReelVaultCore
import com.reelvault.android.data.UploadEndpoint
import com.reelvault.data.models.Tag
import com.reelvault.data.models.Collection as VideoCollection
import com.reelvault.sync.SyncDirection
import com.reelvault.sync.SyncManager
import com.reelvault.sync.SyncProfile
import com.reelvault.sync.SyncRunResult
import kotlinx.coroutines.launch

/**
 * Bottom sheet for starting a catalog sync run.  Shown from [LibraryGridScreen]
 * when the user taps one of the sync toolbar buttons (Local mode + paired server).
 *
 * The sheet is NOT a nav route — it lives entirely in the Compose composition so
 * it doesn't pollute the Android back-stack / nav history.
 *
 * [direction] is pre-set to either [SyncDirection.TO_REMOTE] or
 * [SyncDirection.FROM_REMOTE]; [endpoint] carries the remote server credentials
 * rebuilt from SharedPreferences + the Keychain token (same source as the upload
 * sheet).
 */
@Composable
fun SyncSetupSheet(
    direction: SyncDirection,
    endpoint: UploadEndpoint,
    onDismiss: () -> Unit,
) {
    val scope = rememberCoroutineScope()
    var targetHeight by remember { mutableIntStateOf(1080) }
    var isRunning by remember { mutableStateOf(false) }
    var result by remember { mutableStateOf<SyncRunResult?>(null) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var availableTags by remember { mutableStateOf<List<Tag>>(emptyList()) }
    var availableCollections by remember { mutableStateOf<List<VideoCollection>>(emptyList()) }
    var selectedTagId by remember { mutableStateOf("") }
    var selectedCollectionId by remember { mutableStateOf("") }
    var filterMinRating by remember { mutableIntStateOf(0) }
    var filterColorLabel by remember { mutableStateOf("") }
    var showTagDropdown by remember { mutableStateOf(false) }
    var showCollectionDropdown by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        try {
            val repo = ReelVaultApp.instance.videoRepository
            availableTags = repo.listTags().sortedBy { it.name.lowercase() }
            availableCollections = repo.listCollections()
                .filter { !it.isSmart }
                .sortedBy { it.name.lowercase() }
        } catch (_: Exception) { /* filter options stay empty */ }
    }

    ModalBottomSheet(onDismissRequest = onDismiss) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            Text(
                text = if (direction == SyncDirection.TO_REMOTE) "Sync to Remote" else "Sync from Remote",
                style = MaterialTheme.typography.headlineSmall,
            )
            Spacer(modifier = Modifier.height(8.dp))

            Text(
                text = if (direction == SyncDirection.TO_REMOTE)
                    "Push local videos and catalog data to the paired server"
                else
                    "Pull remote videos to this device in a playable format",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            // Quality selector — only relevant for FROM_REMOTE pulls.
            if (direction == SyncDirection.FROM_REMOTE) {
                Spacer(modifier = Modifier.height(16.dp))
                Text("Target Quality", style = MaterialTheme.typography.labelLarge)
                Spacer(modifier = Modifier.height(8.dp))
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    listOf(720 to "720p", 1080 to "1080p", 0 to "Original").forEach { (h, label) ->
                        FilterChip(
                            selected = targetHeight == h,
                            onClick = { targetHeight = h },
                            label = { Text(label) },
                        )
                    }
                }
            }

            // Filter section
            Spacer(modifier = Modifier.height(16.dp))
            Text("Filter (optional)", style = MaterialTheme.typography.labelLarge)
            Spacer(modifier = Modifier.height(8.dp))

            // Tag dropdown
            ExposedDropdownMenuBox(
                expanded = showTagDropdown,
                onExpandedChange = { showTagDropdown = it },
            ) {
                OutlinedTextField(
                    value = if (selectedTagId.isEmpty()) "Any tag"
                            else availableTags.find { it.id == selectedTagId }?.name ?: "Any tag",
                    onValueChange = {},
                    readOnly = true,
                    label = { Text("Tag") },
                    trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = showTagDropdown) },
                    modifier = Modifier
                        .menuAnchor()
                        .fillMaxWidth(),
                )
                ExposedDropdownMenu(expanded = showTagDropdown, onDismissRequest = { showTagDropdown = false }) {
                    DropdownMenuItem(text = { Text("Any tag") }, onClick = { selectedTagId = ""; showTagDropdown = false })
                    availableTags.forEach { tag ->
                        DropdownMenuItem(
                            text = { Text(tag.name) },
                            onClick = { selectedTagId = tag.id; showTagDropdown = false },
                        )
                    }
                }
            }
            Spacer(modifier = Modifier.height(8.dp))

            // Collection dropdown
            ExposedDropdownMenuBox(
                expanded = showCollectionDropdown,
                onExpandedChange = { showCollectionDropdown = it },
            ) {
                OutlinedTextField(
                    value = if (selectedCollectionId.isEmpty()) "Any collection"
                            else availableCollections.find { it.id == selectedCollectionId }?.name ?: "Any collection",
                    onValueChange = {},
                    readOnly = true,
                    label = { Text("Collection") },
                    trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = showCollectionDropdown) },
                    modifier = Modifier
                        .menuAnchor()
                        .fillMaxWidth(),
                )
                ExposedDropdownMenu(expanded = showCollectionDropdown, onDismissRequest = { showCollectionDropdown = false }) {
                    DropdownMenuItem(text = { Text("Any collection") }, onClick = { selectedCollectionId = ""; showCollectionDropdown = false })
                    availableCollections.forEach { coll ->
                        DropdownMenuItem(
                            text = { Text(coll.name) },
                            onClick = { selectedCollectionId = coll.id; showCollectionDropdown = false },
                        )
                    }
                }
            }
            Spacer(modifier = Modifier.height(8.dp))

            // Rating chips
            Text("Min Rating", style = MaterialTheme.typography.labelMedium)
            Spacer(modifier = Modifier.height(4.dp))
            Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                listOf(0 to "Any", 1 to "★+", 2 to "★★+", 3 to "★★★+", 4 to "★★★★+", 5 to "★★★★★").forEach { (rating, label) ->
                    FilterChip(
                        selected = filterMinRating == rating,
                        onClick = { filterMinRating = rating },
                        label = { Text(label) },
                    )
                }
            }
            Spacer(modifier = Modifier.height(8.dp))

            // Color label chips
            Text("Color Label", style = MaterialTheme.typography.labelMedium)
            Spacer(modifier = Modifier.height(4.dp))
            Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                listOf("" to "Any", "red" to "Red", "yellow" to "Yellow", "green" to "Green", "blue" to "Blue", "purple" to "Purple").forEach { (key, label) ->
                    FilterChip(
                        selected = filterColorLabel == key,
                        onClick = { filterColorLabel = key },
                        label = { Text(label) },
                    )
                }
            }

            // Result summary after a completed run.
            result?.let { r ->
                Spacer(modifier = Modifier.height(16.dp))
                HorizontalDivider()
                Spacer(modifier = Modifier.height(8.dp))
                Text(
                    "${r.completed} synced, ${r.failed} failed",
                    style = MaterialTheme.typography.bodyMedium,
                )
                if (r.errors.isNotEmpty()) {
                    Text(
                        r.errors.first(),
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.error,
                    )
                }
            }

            errorMessage?.let {
                Spacer(modifier = Modifier.height(8.dp))
                Text(
                    it,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.error,
                )
            }

            Spacer(modifier = Modifier.height(24.dp))

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(8.dp, Alignment.End),
            ) {
                TextButton(onClick = onDismiss) { Text("Cancel") }
                Button(
                    onClick = {
                        scope.launch {
                            isRunning = true
                            errorMessage = null
                            result = null
                            try {
                                val localPort = LocalCore.port
                                if (localPort <= 0) {
                                    errorMessage = "Local core is not running"
                                    return@launch
                                }
                                val app = ReelVaultApp.instance
                                val profile = SyncProfile(
                                    name = "Quick Sync",
                                    peerKey = endpoint.fingerprintHex,
                                    direction = direction,
                                    filterJson = buildSyncFilterJson(selectedTagId, selectedCollectionId, filterMinRating, filterColorLabel),
                                    targetHeight = targetHeight,
                                )
                                val manager = SyncManager(
                                    localChannelFactory = app.channelFactory,
                                    localGrpcPort = localPort,
                                    remoteChannelFactory = app.channelFactory,
                                    remoteHost = endpoint.host,
                                    remoteGrpcPort = endpoint.mediaPort - 1,
                                    remoteMediaPort = endpoint.mediaPort,
                                    token = endpoint.token,
                                    fingerprint = endpoint.fingerprintHex,
                                    onIngestSynced = { path, filename, contentHash, derivedHeight ->
                                        ReelVaultCore.nativeIngestSynced(
                                            path, filename, contentHash, derivedHeight,
                                        )
                                    },
                                )
                                manager.startSync(profile)
                                result = manager.lastResult.value
                            } catch (e: Exception) {
                                errorMessage = e.message ?: "Sync failed"
                            } finally {
                                isRunning = false
                            }
                        }
                    },
                    enabled = !isRunning,
                ) {
                    if (isRunning) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(20.dp),
                            strokeWidth = 2.dp,
                        )
                    } else {
                        Text("Start Sync")
                    }
                }
            }

            Spacer(modifier = Modifier.height(16.dp))
        }
    }
}

private fun buildSyncFilterJson(
    tagId: String,
    collectionId: String,
    minRating: Int,
    colorLabel: String,
): String {
    val obj = org.json.JSONObject()
    if (tagId.isNotEmpty()) obj.put("filterTags", org.json.JSONArray().apply { put(tagId) })
    if (collectionId.isNotEmpty()) obj.put("collectionId", collectionId)
    if (minRating > 0) obj.put("filterMinRating", minRating)
    if (colorLabel.isNotEmpty()) obj.put("filterColorLabel", colorLabel)
    return if (obj.length() == 0) "" else obj.toString()
}
